### PR TITLE
Implement existential types

### PR DESCRIPTION
**Pull request by oli-obk**
(not for associated types yet)

r? @nikomatsakis 

cc @Centril @varkor @alexreg

### CODE REVIEW

File: src/librustc_typeck/collect.rs [hunk 1]
```diff
@@ -1149,6 +1168,89 @@ fn type_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
...
+                .typeck_tables_of(def_id)
+                .concrete_existential_types
+                .get(&self.def_id)
+                .cloned();
+            if let Some(ty) = ty {
+                // FIXME(oli-obk): trace the actual span from inference to improve errors
+                let span = self.tcx.def_span(def_id);
+                if let Some((prev_span, prev_ty)) = self.found {
+                    if ty != prev_ty {
+                        // found different concrete types for the existential type
```
Comment (cramertj) [inline src/librustc_typeck/collect.rs]: This won't work for generic existential types because they might be applied with different parameters, right? Is the plan to relax this constraint in the future?

Comment (oli-obk) [reply src/librustc_typeck/collect.rs]: My brain broke when I tried to think about that topic, so I left the problem for tomorrow (literally, not far future). As long as there is only a single defining use, you can use generics already in this PR.

File: src/test/run-pass/existential_type.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+mod boo {
+    pub fn boo() -> super::Boo {
+        "boo"
+    }
+}
+
+// generics
+existential type Cmp<T>: std::cmp::PartialEq<T> + std::fmt::Debug;
+
+fn cmp() -> Cmp<u32> {
```
Comment (cramertj) [inline src/test/run-pass/existential_type.rs]: This should be an error according to the RFC. In order for type parameters to be applied correctly (and fully generally) there needs to be at least one definition site that leaves the parameters fully generic.

e.g.
```rust
existential type Cmp<A, B>;
struct Bar<A, B>(A, B);
fn foo() -> Cmp<u32, u32> {
    Bar(5u32, 6u32)
}
```
What is the definition of `Cmp`? It might seem like it is `type Cmp<A, B> = Bar<A, B>;`, but `type Cmp<A, B> = Cmp<B, A>;` might also be a valid choice.

**Comment by cramertj** [comment]
This is awesome!

File: src/test/run-pass/existential_type.rs (hunk 2 above)
Comment (cramertj) [reply src/test/run-pass/existential_type.rs]: @oli-obk Yes, although it still has to be unified with whatever the "fully defining" usage is to make sure that it *can* work.

**Comment by oli-obk** [comment]
The following cases do not error yet, but should

```rust
existential type Cmp<T>: std::cmp::PartialEq<T>;
//~^ ERROR could not find defining uses

// not a defining use, because it doesn't define *all* possible generics
fn cmp() -> Cmp<u32> {
    5u32 //~ mismatched types
}

existential type Cmp2<T>: std::cmp::PartialEq<T>;
//~^ ERROR could not find defining uses

// not a defining use, because it doesn't define *all* possible generics
// it only defines `Cmp2` for `T: SomeBounds`
fn cmp2<T: std::cmp::PartialEq<T>>(t: T) -> Cmp2<T> {
    t
}
```

**Comment by alexreg** [comment]
Are you going to follow up with associated types in a separate PR, @oli-obk?

**Comment by oli-obk** [comment]
Yea, I want to get the "simple" case working first. I'm slowly getting the hang of generics.

**Comment by alexreg** [comment]
@oli-obk Waiting for review before you fix the above errors? :-)

**Comment by oli-obk** [comment]
I'm working on it

**Comment by oli-obk** [comment]
So uhm, in other news

```rust
trait Bar {}
struct Dummy;
impl Bar for Dummy {}

trait Foo {
    type Assoc: Bar;
    fn foo() -> Self::Assoc;
    fn bar() -> Self::Assoc;
}

existential type Helper: Bar;

impl Foo for i32 {
    type Assoc = Helper;
    fn foo() -> Helper {
        Dummy
    }
    fn bar() -> Helper {
        Dummy
    }
}

fn main() {}
```

works out of the box

File: src/librustc/ty/mod.rs [hunk 3]
```diff
@@ -2857,8 +2863,15 @@ fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     // The param_env of an existential type is its parent's param_env
```
Comment (nikomatsakis) [inline src/librustc/ty/mod.rs]: Can we update this comment? This is specific to `impl Trait`

File: src/librustc_save_analysis/dump_visitor.rs [hunk 4]
```diff
@@ -1238,6 +1238,16 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> DumpVisitor<'l, 'tcx, 'll, O> {
                 // trait.
                 self.visit_ty(ty)
             }
+            ast::ImplItemKind::Existential(ref bounds) => {
+                // FIXME uses of the assoc type should ideally point to this
+                // 'def' and the name here should be a ref to the def in the
+                // trait.
+                for bound in bounds.iter() {
+                    if let ast::GenericBound::Trait(ref trait_ref, _) = *bound {
```
Comment (nikomatsakis) [inline src/librustc_save_analysis/dump_visitor.rs]: Nit: why not get rid of the `*` and the `ref` =)

File: src/librustc_typeck/check/wfcheck.rs [hunk 5]
```diff
@@ -209,6 +210,10 @@ fn check_associated_item<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                     fcx.register_wf_obligation(ty, span, code.clone());
                 }
             }
+            ty::AssociatedKind::Existential => {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/wfcheck.rs]: Hmm. I... don't think this concept even makes sense, does it? I would rather expect that a `existential type` in an impl is more like `-> impl Trait` in a function -- it desugars to the pair of an existential type that is scoped to the impl and defined within. On the **trait** side, this is just a plain associated type.

File: src/librustc_typeck/check/wfcheck.rs [hunk 6]
```diff
@@ -521,7 +532,79 @@ fn check_fn_or_method<'a, 'fcx, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
...
+        fldop: |ty| {
+            if let ty::TyAnon(def_id, substs) = ty.sty {
+                trace!("check_existential_types: anon_ty, {:?}, {:?}", def_id, substs);
+                let anon_node_id = tcx.hir.as_local_node_id(def_id).unwrap();
+                if may_define_existential_type(tcx, fn_def_id, anon_node_id) {
+                    let generics = tcx.generics_of(def_id);
+                    trace!("check_existential_types may define. Generics: {:#?}", generics);
+                    for (subst, param) in substs.iter().zip(&generics.params) {
+                        if let ty::subst::UnpackedKind::Type(ty) = subst.unpack() {
+                            match ty.sty {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/wfcheck.rs]: This isn't quite good enough. We must also check that no type parameters are used *twice*. In particular, `fn foo<T>() -> Foo<T, T>` is wrong.

File: src/librustc_typeck/check/wfcheck.rs [hunk 7]
```diff
@@ -521,7 +532,79 @@ fn check_fn_or_method<'a, 'fcx, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'gcx>,
     // FIXME(#25759) return types should not be implied bounds
     implied_bounds.push(sig.output());
 
-    check_where_clauses(tcx, fcx, span, def_id);
+    check_where_clauses(tcx, fcx, span, def_id, Some(sig.output()));
+}
+
+fn check_existential_types<'a, 'fcx, 'gcx, 'tcx>(
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/wfcheck.rs]: This function should have a comment. I suggest:

---

Checks "defining uses" of existential types to ensure that they meet the restrictions laid for "higher-order pattern unification". This ensures that inference is tractable. In particular, definitions of existential types can only use other generics as arguments, and they cannot repeat an argument. Example:

```rust
existential type Foo<A, B>;

// ok -- `Foo` is applied to two distinct, generic types.
fn a<T, U>() -> Foo<T, U> { .. }

// not ok -- `Foo` is applied to `T` twice.
fn b<T>() -> Foo<T, T> { .. }


// not ok -- `Foo` is applied to a non-generic type.
fn b<T>() -> Foo<T, u32> { .. }
```

Comment (nikomatsakis) [reply src/librustc_typeck/check/wfcheck.rs]: Also, the same applies to lifetimes.

File: src/librustc_typeck/check/writeback.rs [hunk 8]
```diff
@@ -387,11 +388,65 @@ impl<'cx, 'gcx, 'tcx> WritebackCx<'cx, 'gcx, 'tcx> {
...
+                // from being defining
+
+                // Also replace all generic params with the ones from the existential type
+                // definition so
+                // ```rust
+                // existential type Foo<T>: 'static;
+                // fn foo<U>() -> Foo<U> { .. }
+                // ```
+                // figures out the concrete type with `U`, but the stored type is with `T`
+                definition_ty = definition_ty.fold_with(&mut BottomUpFolder {
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/writeback.rs]: Lifetimes?

File: src/librustc_typeck/collect.rs [hunk 9]
```diff
@@ -1149,6 +1168,95 @@ fn type_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     }
 }
 
+fn find_existential_constraints<'a, 'tcx>(
```
Comment (nikomatsakis) [inline src/librustc_typeck/collect.rs]: comment needed -- it'd be great to see some examples of what this code is doing :)

Comment (nikomatsakis) [reply src/librustc_typeck/collect.rs]: ah, I see, is it checking that all uses are compatible?

File: src/test/ui/impl-trait/auto-trait-leak.stderr [hunk 10]
```diff
@@ -1,29 +1,67 @@
-error[E0391]: cycle detected when processing `cycle1`
+error[E0391]: cycle detected when processing `cycle1::{{impl-Trait}}`
```
Comment (nikomatsakis) [inline src/test/ui/impl-trait/auto-trait-leak.stderr]: we should think about making this error better, but that seems orthogonal from this PR

File: src/test/ui/existential_type2.rs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+
+
+#![feature(existential_type)]
+
+fn main() {}
+
+existential type Underconstrained<T: std::fmt::Debug>: 'static;
+//~^ ERROR `U` doesn't implement `std::fmt::Debug`
+
+// not a defining use, because it doesn't define *all* possible generics
```
Comment (nikomatsakis) [inline src/test/ui/existential_type2.rs]: The span of the error here is pretty odd, no? I think we talked about this -- I was *expecting* to see errors arise from the WF checks on this function. I guess perhaps the WF definition for an existential type is not quite right...

File: src/test/ui/existential_type.rs [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+
+#![feature(existential_type)]
+
+fn main() {}
+
+// two definitions with different types
+existential type Foo: std::fmt::Debug;
```
Comment (nikomatsakis) [inline src/test/ui/existential_type.rs]: PS I am resisting complaining about this syntax *so much* 😛

File: src/test/ui/existential_type.rs [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+
+fn foo() -> Foo {
+    ""
+}
+
+fn bar() -> Foo { //~ ERROR defining existential type use differs from previous
+    42i32
+}
+
+// declared but never defined
```
Comment (nikomatsakis) [inline src/test/ui/existential_type.rs]: I think I would rather see these as distinct tests (with suggestive filenames like `src/test/ui/existential_type/declared_but_never_defined.rs`). These "mega tests" files always make it kind of hard for me to sort of which parts belong to which test.

File: src/test/ui/existential_type.rs [hunk 14]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
```
Comment (nikomatsakis) [inline src/test/ui/existential_type.rs]: please put these tests into `src/test/ui/existential_type/XXX.rs` files

File: src/test/ui/existential_type.rs [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+// declared but never defined
+existential type Bar: std::fmt::Debug; //~ ERROR could not find defining uses
+
+mod boo {
+    // declared in module but not defined inside of it
+    pub existential type Boo: ::std::fmt::Debug; //~ ERROR could not find defining uses
+}
+
+fn bomp() -> boo::Boo {
+    "" //~ ERROR mismatched types
```
Comment (nikomatsakis) [inline src/test/ui/existential_type.rs]: What would happen if this were to be `loop { }` or `panic!()`... no error?

File: src/test/ui/existential_type2.rs (hunk 11 above)
Comment (nikomatsakis) [reply src/test/ui/existential_type2.rs]: Yep, this code doesn't look right:

https://github.com/rust-lang/rust/blob/3d5753fda1ee8f729da1061e931e13b043f479a5/src/librustc/ty/wf.rs#L363-L367

We probably want to rewrite this to basically do the same as structs/enums:

https://github.com/rust-lang/rust/blob/3d5753fda1ee8f729da1061e931e13b043f479a5/src/librustc/ty/wf.rs#L292-L296

We *might* want to only do that for the case of explicit `existential type` declarations -- though I imagine it will be fine either way.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Sorry for being slow. This is so awesome. Left a bunch of minor comments below.

File: src/librustc_typeck/check/wfcheck.rs (hunk 5 above)
Comment (oli-obk) [reply src/librustc_typeck/check/wfcheck.rs]: yea, I was just following the RFC, but all of the associated existential code is stubbed out anyway. I'll remove it from this PR entirely

Comment (nikomatsakis) [reply src/librustc_typeck/check/wfcheck.rs]: yes; my point is that even if we allow the syntax I wouldn't necessarily expect it to show up at the `ty` level (though perhaps it will wind up being necessary for...reasons)

File: src/test/ui/existential_type.rs (hunk 15 above)
Comment (oli-obk) [reply src/test/ui/existential_type.rs]: Added tests.

Non-defining uses are allowed to be `!`, but with defining uses you get a type mismatch error. If you wrote `if foo { loop {} } else { "" }` then it would work again.

File: src/test/ui/existential_type.rs (hunk 14 above)
Comment (oli-obk) [reply src/test/ui/existential_type.rs]: done

File: src/test/ui/existential_type.rs (hunk 13 above)
Comment (oli-obk) [reply src/test/ui/existential_type.rs]: done

File: src/librustc_save_analysis/dump_visitor.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc_save_analysis/dump_visitor.rs]: we should make a clippy lint for that

File: src/test/ui/existential_type2.rs (hunk 11 above)
Comment (oli-obk) [reply src/test/ui/existential_type2.rs]: Yea we need to skip it for `impl Trait`, otherwise https://github.com/rust-lang/rust/blob/master/src/test/ui/nll/ty-outlives/impl-trait-outlives.rs passes without errors

Comment (oli-obk) [reply src/test/ui/existential_type2.rs]: Also

```rust
existential type WrongGeneric<T>: 'static;
//~^ ERROR the parameter type `T` may not live long enough

fn wrong_generic<T>(t: T) -> WrongGeneric<T> {
    t
}
```

stops being an error with NLL if I do that change.

**Comment by nikomatsakis** [comment]
OK, so @oli-obk and I were talking about some of the suboptimal errors -- these are a symptom of the current setup, where we check for consistency of the inferred values for an existential type during the "collect" query. The problem is that this invokes the "check fn body" queries which never wind up running the "wfcheck" query on those fns. If we force the "check fn body" query to invoke "wfcheck" (which it really *ought* to do), then I see the errors I expect. We decided to try and do this more generally in a follow-up.

Therefore, r=me @oli-obk post rebase. 

Diff:

```diff
diff --git a/src/librustc_typeck/check/mod.rs b/src/librustc_typeck/check/mod.rs
index 3e81262424..3a81891624 100644
--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -826,6 +826,13 @@ fn typeck_tables_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
     let id = tcx.hir.as_local_node_id(def_id).unwrap();
     let span = tcx.hir.span(id);

+    // For the base function, first run the well-formedness check.
+    // This is a hack -- there should just be a `check_well_formed`
+    // query that works for any def-id.
+    if let Node::NodeItem(_) = tcx.hir.get(id) {
+        ty::query::queries::check_item_well_formed::ensure(tcx, def_id);
+    }
+
     // Figure out what primary body this item has.
     let (body_id, fn_decl) = primary_body_of(tcx, id).unwrap_or_else(|| {
         span_bug!(span, "can't type-check body of {:?}", def_id);
```

**Comment by oli-obk** [comment]
@bors r=nikomatsakis

Finally figured out how to run this locally. Passes now

**Comment by kennytm** [comment]
This breakage is inconveniently timed...

```
[01:02:31] error[E0004]: non-exhaustive patterns: `AssociatedExistential(_)` not covered
[01:02:31]    --> tools/clippy/clippy_lints/src/utils/mod.rs:964:11
[01:02:31]     |
[01:02:31] 964 |     match def {
[01:02:31]     |           ^^^ pattern `AssociatedExistential(_)` not covered
[01:02:31]
[01:02:31] error[E0004]: non-exhaustive patterns: `Existential(_)` not covered
[01:02:31]   --> tools/clippy/clippy_lints/src/utils/inspector.rs:66:15
[01:02:31]    |
[01:02:31] 66 |         match item.node {
[01:02:31]    |               ^^^^^^^^^ pattern `Existential(_)` not covered
[01:02:31]
[01:02:31] error[E0004]: non-exhaustive patterns: `Existential(_)` not covered
[01:02:31]    --> tools/clippy/clippy_lints/src/missing_doc.rs:176:26
[01:02:31]     |
[01:02:31] 176 |         let desc = match impl_item.node {
[01:02:31]     |                          ^^^^^^^^^^^^^^ pattern `Existential(_)` not covered
[01:02:31]
[01:02:31] error[E0004]: non-exhaustive patterns: `Existential(_)` not covered
[01:02:31]    --> tools/clippy/clippy_lints/src/missing_inline.rs:165:26
[01:02:31]     |
[01:02:31] 165 |         let desc = match impl_item.node {
[01:02:31]     |                          ^^^^^^^^^^^^^^ pattern `Existential(_)` not covered
[01:02:31]
[01:02:31] error: aborting due to 4 previous errors
[01:02:31]
[01:02:31] For more information about this error, try `rustc --explain E0004`.
[01:02:31] error: Could not compile `clippy_lints`.
```

**Comment by alexreg** [comment]
@kennytm Fixed. https://github.com/rust-lang-nursery/rust-clippy/pull/2939
